File: fino/src/types/calibration_item.jl
using Dates
using Random

using .DividendManager: Dividend

struct CalibrationItem
    mny_fwd_ln::Vector{Float64}
    strike::Vector{Float64}
    calculation_date::Date
    tenor_dt::Date
    v_is_call::Vector{UInt8}
    iv::Vector{Float64}
    price::Vector{Float64}
    spot::Vector{Float64}
    dividends::Vector{Dividend}
    weights::Union{Vector{Float64}, Nothing}
    vega::Union{Vector{Float64}, Nothing}
    ts::Union{Vector{DateTime}, Nothing}
    bid_price::Vector{Float64}
    ask_price::Vector{Float64}
    dividend_yield::Union{Float64, Nothing}  # not needed anymore for American options
end

function CalibrationItem(;
    mny_fwd_ln::Vector,
    strike::Vector,
    calculation_date::Date,
    tenor_dt::Date,
    v_is_call::Vector{UInt8},
    iv::Vector,
    price::Vector,
    spot::Vector,
    dividends::Vector{Dividend}=Dividend[],
    weights::Union{Vector, Nothing}=nothing,
    vega::Union{Vector, Nothing}=nothing,
    ts::Union{Vector{DateTime}, Nothing}=nothing,
    bid_price::Union{Vector, Nothing}=nothing,
    ask_price::Union{Vector, Nothing}=nothing,
    dividend_yield::Union{Float64, Nothing}=nothing
)
    n = length(price)
    bid_price_fixed = bid_price === nothing ? fill(NaN, n) : bid_price
    ask_price_fixed = ask_price === nothing ? fill(NaN, n) : ask_price
    
    return CalibrationItem(
        mny_fwd_ln, strike, calculation_date, tenor_dt, v_is_call, iv, price, spot,
        dividends, weights, vega, ts, bid_price_fixed, ask_price_fixed, dividend_yield
    )
end

function Base.getproperty(item::CalibrationItem, sym::Symbol)
    if sym === :tenor
        return get_tenor(item.tenor_dt, item.calculation_date)
    end
    return getfield(item, sym)
end

function Base.:(+)(a::CalibrationItem, b::CalibrationItem)
    if a.tenor_dt != b.tenor_dt
        throw(ArgumentError("Cannot add CalibrationItems with different tenor_dt"))
    end
    if a.calculation_date != b.calculation_date
        throw(ArgumentError("Cannot add CalibrationItems with different calculation_date"))
    end

    return CalibrationItem(
        mny_fwd_ln=vcat(a.mny_fwd_ln, b.mny_fwd_ln),
        strike=vcat(a.strike, b.strike),
        calculation_date=a.calculation_date,
        tenor_dt=a.tenor_dt,
        v_is_call=vcat(a.v_is_call, b.v_is_call),
        iv=vcat(a.iv, b.iv),
        price=vcat(a.price, b.price),
        spot=vcat(a.spot, b.spot),
        dividends=a.dividends,
        weights=a.weights === nothing || b.weights === nothing ? nothing : vcat(a.weights, b.weights),
        vega=a.vega === nothing || b.vega === nothing ? nothing : vcat(a.vega, b.vega),
        ts=a.ts === nothing || b.ts === nothing ? nothing : vcat(a.ts, b.ts),
        bid_price=vcat(a.bid_price, b.bid_price),
        ask_price=vcat(a.ask_price, b.ask_price),
        dividend_yield=a.dividend_yield
    )
end

function downsample(item::CalibrationItem, n::Int)
    n_actual = min(n, length(item.price))
    ix_sample = randperm(length(item.price))[1:n_actual]
    
    return CalibrationItem(
        mny_fwd_ln=item.mny_fwd_ln[ix_sample],
        strike=item.strike[ix_sample],
        calculation_date=item.calculation_date,
        tenor_dt=item.tenor_dt,
        v_is_call=item.v_is_call[ix_sample],
        iv=item.iv[ix_sample],
        price=item.price[ix_sample],
        spot=item.spot[ix_sample],
        dividends=item.dividends,
        weights=item.weights === nothing ? nothing : item.weights[ix_sample],
        vega=item.vega === nothing ? nothing : item.vega[ix_sample],
        ts=item.ts === nothing ? nothing : item.ts[ix_sample],
        bid_price=item.bid_price[ix_sample],
        ask_price=item.ask_price[ix_sample],
        dividend_yield=item.dividend_yield
    )
end

function df2calibration_items(df_in::DataFrame, calc_date::Date, iv_col_nm, price_col_nm, spot_col_nm, dividends::Vector{Dividend};
    weight_col_nm="vega_mid_price_iv", vega_col_nm=nothing, max_samples_by_tenor=nothing, seed=1234)
    
    # Filter out NaN values
    df = df_in[.!ismissing.(df_in[!, price_col_nm]) .& .!ismissing.(df_in[!, iv_col_nm]), :]
    
    if weight_col_nm !== nothing
        df = df[(.!ismissing.(df[!, weight_col_nm])) .& (df[!, weight_col_nm] .> 0), :]
    end
    
    Random.seed!(seed)
    calibration_items = CalibrationItem[]
    
    # Group by expiry
    for expiry in unique(df[!, :expiry])
        s_df = df[df[!, :expiry] .== expiry, :]
        
        if max_samples_by_tenor !== nothing
            n_samples = min(max_samples_by_tenor, nrow(s_df))
            ix_sample = randperm(nrow(s_df))[1:n_samples]
            df_sample = s_df[ix_sample, :]
        else
            df_sample = s_df
        end
        
        push!(calibration_items, CalibrationItem(
            mny_fwd_ln=Float64.(df_sample[!, :moneyness_fwd_ln]),
            strike=Float64.(df_sample[!, :strike]),
            calculation_date=calc_date,
            tenor_dt=expiry,
            v_is_call=Bool.(df_sample[!, :v_is_call]),
            iv=Float64.(df_sample[!, iv_col_nm]),
            price=Float64.(df_sample[!, price_col_nm]),
            spot=Float64.(df_sample[!, spot_col_nm]),
            dividend_yield=0.0,
            dividends=dividends,
            weights=weight_col_nm !== nothing ? Float64.(df_sample[!, weight_col_nm]) : nothing,
            vega=vega_col_nm !== nothing ? Float64.(df_sample[!, vega_col_nm]) : nothing,
            ts=DateTime.(df_sample[!, :ts]),
            bid_price="bid_close" in names(df_in) ? Float64.(df_sample[!, :bid_close]) : nothing,
            ask_price="ask_close" in names(df_in) ? Float64.(df_sample[!, :ask_close]) : nothing,
        ))
    end
    
    return calibration_items
end

function union_calibration_items(ci::Vector{CalibrationItem}...)
    dct_ci = Dict{Date, CalibrationItem}()
    
    for c in Iterators.flatten(ci)
        key = c.tenor_dt
        if !haskey(dct_ci, key)
            dct_ci[key] = c
        else
            dct_ci[key] = dct_ci[key] + c
        end
    end
    
    return sort(collect(values(dct_ci)), by=x -> x.tenor_dt)
end

File: options/types/calibration_item.py
import numpy as np
import pandas as pd

from itertools import chain
from dataclasses import dataclass
from datetime import date, datetime
from typing import List, Dict
from options.helper import get_tenor
from options.types.dividend import Dividend


@dataclass
class CalibrationItem:
    mny_fwd_ln: np.ndarray
    strike: np.ndarray
    calculation_date: date
    tenor_dt: date
    rights: np.ndarray
    iv: np.ndarray
    price: np.ndarray
    spot: np.ndarray
    dividends: List[Dividend] = ()
    weights: np.ndarray = None
    vega: np.ndarray = None
    ts: np.ndarray[datetime] = None
    bid_price: np.ndarray = None
    ask_price: np.ndarray = None
    dividend_yield: float = None  # not needed anymore for American options.

    def __post_init__(self):
        if self.bid_price is None:
            self.bid_price = np.empty_like(self.price)
            self.bid_price[:] = np.nan
        if self.ask_price is None:
            self.ask_price = np.empty_like(self.price)
            self.ask_price[:] = np.nan

    @property
    def tenor(self):
        return get_tenor(self.tenor_dt, self.calculation_date)

    def __add__(self, other):
        if not isinstance(other, CalibrationItem):
            raise NotImplemented
        if self.tenor_dt != other.tenor_dt:
            raise NotImplemented
        if self.calculation_date != other.calculation_date:
            raise NotImplemented

        mny_fwd_ln = np.concatenate([self.mny_fwd_ln, other.mny_fwd_ln])
        strike = np.concatenate([self.strike, other.strike])
        rights = np.concatenate([self.rights, other.rights])
        iv = np.concatenate([self.iv, other.iv])
        price = np.concatenate([self.price, other.price])
        spot = np.concatenate([self.spot, other.spot])
        weights = np.concatenate([self.weights, other.weights]) if self.weights is not None else None
        vega = np.concatenate([self.vega, other.vega]) if self.vega is not None else None
        ts = np.concatenate([self.ts, other.ts]) if self.ts is not None else None
        bid_price = np.concatenate([self.bid_price, other.bid_price])
        ask_price = np.concatenate([self.ask_price, other.ask_price])

        return CalibrationItem(
            mny_fwd_ln=mny_fwd_ln,
            strike=strike,
            calculation_date=self.calculation_date,
            tenor_dt=self.tenor_dt,
            rights=rights,
            iv=iv,
            price=price,
            spot=spot,
            dividends=self.dividends,
            weights=weights,
            vega=vega,
            ts=ts,
            bid_price=bid_price,
            ask_price=ask_price
        )

    def downsample(self, n: int) -> 'CalibrationItem':
        ix_all = np.arange(len(self.price))
        ix_sample = np.random.choice(ix_all, size=n, replace=False)
        return CalibrationItem(
            mny_fwd_ln=self.mny_fwd_ln[ix_sample],
            strike=self.strike[ix_sample],
            calculation_date=self.calculation_date,
            tenor_dt=self.tenor_dt,
            rights=self.rights[ix_sample],
            iv=self.iv[ix_sample],
            price=self.price[ix_sample],
            spot=self.spot[ix_sample],
            dividends=self.dividends,
            weights=self.weights[ix_sample] if self.weights is not None else None,
            vega=self.vega[ix_sample] if self.vega is not None else None,
            ts=self.ts[ix_sample],
            bid_price=self.bid_price[ix_sample],
            ask_price=self.ask_price[ix_sample],
        )


def df2calibration_items(df_in: pd.DataFrame, calc_date: date, iv_col_nm, price_col_nm, spot_col_nm, dividends: List[Dividend], weight_col_nm='vega_mid_price_iv', vega_col_nm=None, max_samples_by_tenor=None, seed=1234) -> List[CalibrationItem]:
    """2) Refactor to include quotes, at least during plotting."""
    df = df_in[df_in[price_col_nm].notna() & df_in[iv_col_nm].notna()]
    if weight_col_nm:
        df = df[(df[weight_col_nm].notna()) & (df[weight_col_nm] > 0)]
    calibration_items = []
    for expiry, s_df in df.groupby('expiry'):
        df_sample = s_df.sample(n=min(max_samples_by_tenor, len(s_df)), random_state=seed) if max_samples_by_tenor else s_df
        calibration_items.append(
            CalibrationItem(
                mny_fwd_ln=df_sample['moneyness_fwd_ln'].values.astype(float),
                strike=np.array(df_sample.index.get_level_values('strike').astype(float)),
                calculation_date=calc_date,
                tenor_dt=expiry,
                rights=np.array(df_sample.index.get_level_values('right').astype(str)),
                iv=df_sample[iv_col_nm].values.astype(float),
                price=df_sample[price_col_nm].values.astype(float),
                spot=df_sample[spot_col_nm].values.astype(float),
                dividend_yield=0,
                dividends=dividends,
                weights=df_sample[weight_col_nm].values.astype(float) if weight_col_nm else None,
                vega=df_sample[vega_col_nm].values.astype(float) if vega_col_nm else None,
                ts=pd.to_datetime(df_sample.index.get_level_values('ts')),
                bid_price=df_sample['bid_close'].values if 'bid_close' in df_in.columns else None,
                ask_price=df_sample['ask_close'].values if 'ask_close' in df_in.columns else None,
            ))
    return calibration_items


def union_calibration_items(*ci: List[CalibrationItem]) -> List[CalibrationItem]:
    """Unions quotes and trades into a single calibration item. Grouped by tenor_dt"""
    dct_ci: Dict[date, CalibrationItem] = {}
    for c in chain(*ci):
        key = c.tenor_dt
        if key not in dct_ci:
            dct_ci[key] = c
        else:
            dct_ci[key] += c
    return sorted(dct_ci.values(), key=lambda x: x.tenor_dt)
